File: src/ZoneCodeGenerator/Interface/Arguments/CommandLineOption.cs
using System.Collections.Generic;

namespace ZoneCodeGenerator.Interface.Arguments
{
    class CommandLineOption
    {
        public string ShortName { get; private set; }
        public string LongName { get; private set; }
        public string Description { get; private set; }
        public string Category { get; private set; }
        public bool MultiUse { get; private set; }

        private readonly List<string> parameters;
        public IReadOnlyList<string> Parameters => parameters.AsReadOnly();
        public int ParameterCount => parameters.Count;

        private CommandLineOption()
        {
            ShortName = null;
            LongName = null;
            Description = "";
            Category = "";
            MultiUse = false;
            parameters = new List<string>();
        }

        public class CommandLineOptionBuilder
        {
            private readonly CommandLineOption option;

            private CommandLineOptionBuilder()
            {
                option = new CommandLineOption();
            }

            public static CommandLineOptionBuilder Create()
            {
                return new CommandLineOptionBuilder();
            }

            public CommandLineOptionBuilder WithShortName(string shortName)
            {
                option.ShortName = shortName;
                return this;
            }

            public CommandLineOptionBuilder WithLongName(string longName)
            {
                option.LongName = longName;
                return this;
            }

            public CommandLineOptionBuilder WithDescription(string description)
            {
                option.Description = description;
                return this;
            }

            public CommandLineOptionBuilder WithCategory(string category)
            {
                option.Category = category;
                return this;
            }

            public CommandLineOptionBuilder WithParameter(string argumentName)
            {
                option.parameters.Add(argumentName);
                return this;
            }

            public CommandLineOptionBuilder Reusable()
            {
                option.MultiUse = true;
                return this;
            }

            public CommandLineOption Build()
            {
                return option;
            }
        }
    }
}

File: src/Utils/Utils/Arguments/CommandLineOption.cpp
#include "CommandLineOption.h"

CommandLineOption::CommandLineOption()
{
    m_short_name = "";
    m_long_name = "";
    m_description = "";
    m_category = "";
    m_multi_use = false;
}

CommandLineOption::Builder& CommandLineOption::Builder::Create()
{
    return *new Builder();
}

CommandLineOption::Builder& CommandLineOption::Builder::WithShortName(std::string shortName)
{
    m_option.m_short_name = std::move(shortName);
    return *this;
}

CommandLineOption::Builder& CommandLineOption::Builder::WithLongName(std::string longName)
{
    m_option.m_long_name = std::move(longName);
    return *this;
}

CommandLineOption::Builder& CommandLineOption::Builder::WithDescription(std::string description)
{
    m_option.m_description = std::move(description);
    return *this;
}

CommandLineOption::Builder& CommandLineOption::Builder::WithCategory(std::string category)
{
    m_option.m_category = std::move(category);
    return *this;
}

CommandLineOption::Builder& CommandLineOption::Builder::WithParameter(std::string parameterName)
{
    m_option.m_parameters.push_back(std::move(parameterName));
    return *this;
}

CommandLineOption::Builder& CommandLineOption::Builder::Reusable()
{
    m_option.m_multi_use = true;
    return *this;
}

CommandLineOption* CommandLineOption::Builder::Build() const
{
    auto* result = new CommandLineOption(m_option);
    delete this;

    return result;
}
